nucypher/blockchain/eth/sol/source/contracts/PREApplication.sol

```diff
@@ -0,0 +1,652 @@
+            _token.totalSupply() > 0,
+            "Wrong input parameters"
+        );
+        rewardDuration = _rewardDuration;
+        deauthorizationDuration = _deauthorizationDuration;
+        minAuthorization = _minAuthorization;
+        token = _token;
+        tStaking = _tStaking;
+        minOperatorSeconds = _minOperatorSeconds;
+    }
+
+    /**
+    * @dev Update reward for the specified staking provider
+    */
+    modifier updateReward(address _stakingProvider) {
+        updateRewardInternal(_stakingProvider);
+        _;
+    }
+
+    /**
+    * @dev Checks caller is T staking contract
+    */
+    modifier onlyStakingContract()
+    {
+        require(msg.sender == address(tStaking), "Caller must be the T staking contract");
+        _;
+    }
+
+    /**
+    * @dev Checks caller is a staking provider or stake owner
+    */
+    modifier onlyOwnerOrStakingProvider(address _stakingProvider)
+    {
+        require(isAuthorized(_stakingProvider), "Not owner or provider");
+        if (_stakingProvider != msg.sender) {
+            (address owner,,) = tStaking.rolesOf(_stakingProvider);
+            require(owner == msg.sender, "Not owner or provider");
+        }
+        _;
+    }
+
+    /**
+     * @notice Initialize function for using with OpenZeppelin proxy
+     */
+    function initialize() external initializer {
+        __Ownable_init();
+    }
+
+    //------------------------Reward------------------------------
+
+    /**
+    * @notice Set reward distributor address
+    */
+    function setRewardDistributor(address _rewardDistributor)
+        external
+        onlyOwner
+    {
+        rewardDistributor = _rewardDistributor;
+        emit RewardDistributorSet(_rewardDistributor);
+    }
+
+    /**
+    * @notice Update reward for the specified staking provider
+    * @param _stakingProvider Staking provider address
+    */
+    function updateRewardInternal(address _stakingProvider) internal {
+        rewardPerTokenStored = rewardPerToken();
+        lastUpdateTime = lastTimeRewardApplicable();
+        if (_stakingProvider != address(0)) {
+            StakingProviderInfo storage info = stakingProviderInfo[_stakingProvider];
+            info.tReward = earned(_stakingProvider);
+            info.rewardPerTokenPaid = rewardPerTokenStored;
+        }
+    }
+
+    /**
+    * @notice Returns last time when reward was applicable
+    */
+    function lastTimeRewardApplicable() public view returns (uint256) {
+        return Math.min(block.timestamp, periodFinish);
+    }
+
+    /**
+    * @notice Returns current value of reward per token
+    */
+    function rewardPerToken() public view returns (uint96) {
+        if (authorizedOverall == 0) {
+            return rewardPerTokenStored;
+        }
+        uint256 result = rewardPerTokenStored +
+                (lastTimeRewardApplicable() - lastUpdateTime)
+                * rewardRate
+                * 1e18
```

Comment:
Why is this per token, and not a smaller unit?

Reply:
This part taken from unipool without changing, so I think to decrease chance of overflow in `pushReward`. I guess we can use smaller units if we want